### PR TITLE
Add support for Razer Blade Advanced 15 Mid 2021

### LINKED ISSUE #1542: Support for Razer Blade 15 Advanced (Mid 2021)
### Device Info

* Name: Razer Blade 15 Advanced - Mid 2021 
* Product Number: RZ09-0409BWD3
* Model Number: RZ09-0409

### Device Information
```
lsusb -d 1532: -v

Bus 003 Device 003: ID 1532:0276 Razer USA, Ltd Razer Blade
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0276 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)

```

```
lrwxrwxrwx 1 root root   9 Jun 29 10:54 usb-Azurewave_Integrated_RGB_Camera_01.00.00-event-if00 -> ../event4
lrwxrwxrwx 1 root root   9 Jun 29 10:54 usb-Azurewave_Integrated_RGB_Camera_01.00.00-event-if02 -> ../event5
lrwxrwxrwx 1 root root  10 Jun 29 10:54 usb-Razer_Razer_Blade-event-if01 -> ../event10
lrwxrwxrwx 1 root root   9 Jun 29 10:54 usb-Razer_Razer_Blade-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 Jun 29 10:54 usb-Razer_Razer_Blade-if01-event-kbd -> ../event7
lrwxrwxrwx 1 root root  10 Jun 29 10:54 usb-Razer_Razer_Blade-if02-event-mouse -> ../event11
lrwxrwxrwx 1 root root   9 Jun 29 10:54 usb-Razer_Razer_Blade-if02-mouse -> ../mouse2

```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

### PR DESCRIPTION
**Pull request by j-maynard**
This is an initial commit to support the Razer Blade Advanced 15 Mid 2021 release.  This PR was tested on the 4k model of the Blade Advanced and I can confirm changing keyboard color and most profiles other than Startlight (random) seem to work.  I can't find the references to the logo to allow logo control to work.  If someone can point me in the right direction I might be able to fix that.

Fixes #1542

### CODE REVIEW

**Comment by j-maynard** [comment]
Resolves #1542

File: README.md [hunk 1]
```diff
@@ -68,6 +68,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade Stealth (Early 2020)              |  1532:0252  |
 | Razer Blade 15 Advanced (2020)                |  1532:0253  |
 | Razer Blade 15 (Early 2020) Base              |  1532:0255  |
+| Razer Blade 15 Advanced (Mid 2021)            |  1532:0276  |
```
Comment (z3ntu) [inline README.md]: Please put the device after `Razer Blade 15 Advanced (Early 2021)` so it's sorted by usb pid

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 2]
```diff
@@ -1304,6 +1304,25 @@ class RazerBlade15Advanced2020(_RippleKeyboard):
     DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1651/razer-blade-15-advanced-2020.png"
 
 
+class RazerBlade15Advanced2021(_RippleKeyboard):
+    """
+    Class for the Razer Blade 15 Advanced Mid (2021)
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: `Mid (2021)` -> `(Mid 2021)`

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -70,6 +70,7 @@
 #define USB_DEVICE_ID_RAZER_BLADE_STUDIO_EDITION_2019 0x024D
 #define USB_DEVICE_ID_RAZER_BLADE_STEALTH_EARLY_2020 0x0252
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_2020 0x0253
+#define USB_DEVICE_ID_RAZER_BLADE_15_ADV_MID_2021 0x0276
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: please move down below USB_DEVICE_ID_RAZER_BLADE_15_ADV_EARLY_2021 for sorting

File: install_files/udev/99-razer.rules [hunk 4]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|026a|026d|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0259|025d|025e|026a|026d|0a24|0276", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: move before 0a24 so it's sorted by pid

**Review by z3ntu — CHANGES_REQUESTED** [review]
A couple of minor comments

Regarding logo control, you have added 'get_logo_active' & 'set_logo_active' which are the functions used for controlling the logo. I'm currently not sure which frontends support this, but you can also open e.g. d-feet , session bus, org.razer, razer.device.lighting.logo setLogoActive and try it manually there.

**Comment by j-maynard** [comment]
> Regarding logo control, you have added 'get_logo_active' & 'set_logo_active' which are the functions used for controlling the logo. I'm currently not sure which frontends support this, but you can also open e.g. d-feet , session bus, org.razer, razer.device.lighting.logo setLogoActive and try it manually there.

Thanks for the review... I've tested the logo and its all working and I've made the changes you asked for.  I only ask about the logo because Synapse has the option for a breathing logo.  I'm not sure if this is a new feature or a software implementation.

**Comment by z3ntu** [comment]
Interesting. Would be interesting to have USB packet captures of Synapse doing that. Currently our code only has support for on/off for the logo I believe.

**Comment by julian9499** [comment]
Should that still be added or would this pull request suffice for now?

**Comment by j-maynard** [comment]
I think if this pull request is all good we should maybe merge it and then take a look at the breathing logo effect.  I need to work out how to do packet capture for the Razer Thumderbolt Dock as well.  I can raise a seperate issue for the breathing logo so we can capture that as a thing.

Issue raised at https://github.com/openrazer/openrazer/issues/1574 if someone can label it as requiring a pcap.  Also I'll check on last years Blade Advanced 4k to see if it has the same option but I won't be able to check that for another week as I am away from home.

**Comment by julian9499** [comment]
Yeah same thought although I have 0 experience with USB packets capturing I might look into the inner workings of the breathing logo a bit later.

**Comment by julian9499** [comment]
@z3ntu would this pull request suffice for now then?

**Review by z3ntu — APPROVED** [review]
LGTM, thank you for the pull request!

**Pull request merged**